File: katas/5 kyu/the-hashtag-generator/solution.py
# the-hashtag-generator
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
def generate_hashtag(s):
  s = '#'+s.title()
  s = ''.join(s.split())
  if len(s) in range(2,140):
    return s
  else:    
    return False
# ------------------------------ CLEVER SOLUTION ----------------------------- #
def generate_hashtag(s):
  output = "#"
  
  for word in s.split():
      output += word.capitalize()
  
  return False if (len(s) == 0 or len(output) > 140) else output

def generate_hashtag(s):
  ans = '#'+ str(s.title().replace(' ',''))
  return s and not len(ans)>140 and ans or False
# ----------------------------------- TEST ----------------------------------- #
test.describe("Basic tests")
test.assert_equals(generate_hashtag(''), False, 'Expected an empty string to return False')
test.assert_equals(generate_hashtag('Do We have A Hashtag')[0], '#', 'Expeted a Hashtag (#) at the beginning.')
test.assert_equals(generate_hashtag('Codewars'), '#Codewars', 'Should handle a single word.')
test.assert_equals(generate_hashtag('Codewars      '), '#Codewars', 'Should handle trailing whitespace.')
test.assert_equals(generate_hashtag('Codewars Is Nice'), '#CodewarsIsNice', 'Should remove spaces.')
test.assert_equals(generate_hashtag('codewars is nice'), '#CodewarsIsNice', 'Should capitalize first letters of words.')
test.assert_equals(generate_hashtag('CodeWars is nice'), '#CodewarsIsNice', 'Should capitalize all letters of words - all lower case but the first.')
test.assert_equals(generate_hashtag('c i n'), '#CIN', 'Should capitalize first letters of words even when single letters.')
test.assert_equals(generate_hashtag('codewars  is  nice'), '#CodewarsIsNice', 'Should deal with unnecessary middle spaces.')
test.assert_equals(generate_hashtag('Looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong Cat'), False, 'Should return False if the final word is longer than 140 chars.')

File: katas/5 kyu/the-hashtag-generator/solution.php
<!-- the-hashtag-generator -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function generateHashtag($str) {
  $str = preg_replace('/\s+/', ' ', $str);
  if ($str == "" || $str == " " || strlen($str) >= 140) return false;
  $str = "#" . ucwords($str);
  $str = implode("" , explode(" " , $str));
  return $str;
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
function generateHashtag($str) {
  
  $str = str_replace(' ', '', ucwords($str));
  
  return (strlen($str) > 140 || strlen($str) == 1) ? false : "#" .  $str;
}
/* ---------------------------------- TEST ---------------------------------- */
class MyTestCases extends TestCase
{
    public function testThatSomethingShouldHappen() {
      $this->assertSame(false, generateHashtag(''), 'Expected an empty string to return false');
      $this->assertSame(false, generateHashtag(str_repeat(' ', 200)), "Still an empty string");
      $this->assertSame('#Codewars', generateHashtag('Codewars'), 'Should handle a single word and add a hashtag at the beginning.');
      $this->assertSame('#Codewars', generateHashtag('Codewars      '), 'Should handle trailing whitespace.');
      $this->assertSame('#CodewarsIsNice', generateHashtag('Codewars Is Nice'), 'Should remove spaces.');
      $this->assertSame('#CodewarsIsNice', generateHashtag('codewars is nice'), 'Should capitalize first letters of words.');
      $this->assertSame('#CodeWars', generateHashtag('Code' . str_repeat(' ', 140) . 'wars'));
      $this->assertSame(false, generateHashtag('Looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong Cat'), 'Should return false if the final word is longer than 140 chars.');
      $this->assertSame("#A" . str_repeat("a", 138), generateHashtag(str_repeat("a", 139)), "Should work");
      $this->assertSame(false, generateHashtag(str_repeat("a", 140)), "Too long");
    }
}
